### PR TITLE
Push actors outside of the to-place footprint, if possible.

### LINKED ISSUE #3682: Smart Infantry Units : they could move to let the player install a building [$30]
What about infantry units that automatically move if the player tries to install a building where they are standing ?

It would just be a short move, just to let the building being placed.

<bountysource-plugin>
---
There is a **[$30 open bounty](https://www.bountysource.com/issues/792109-smart-infantry-units-they-could-move-to-let-the-player-install-a-building-15?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by phrohdoh**
ra
d2k
ts
ra2

Closes #3682.

### CODE REVIEW

**Comment by Phrohdoh** [comment]
Fixed the incorrect using directive ordering (which caused Travis to fail).

File: OpenRA.Mods.Common/Traits/Player/PlaceBuilding.cs [hunk 1]
```diff
@@ -114,5 +123,32 @@ static int GetNumBuildables(Player p)
...
+		public static void TryMoveBlockers(Actor self, string toPlaceActorName, BuildingInfo buildingInfo, CPos topLeft)
+		{
+			var world = self.World;
+			var rules = world.Map.Rules;
+			var tiles = FootprintUtils.Tiles(rules, toPlaceActorName, buildingInfo, topLeft);
+			var perimeter = Util.ExpandFootprint(tiles, true).Except(tiles).ToList();
+
+			foreach (var tile in perimeter)
+				if (!world.Map.Contains(tile))
+					perimeter.Remove(tile);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Player/PlaceBuilding.cs]: This can be folded into the linq query above.

**Comment by Phrohdoh** [comment]
Updated per @pchote's request.

**Comment by chrisforbes** [comment]
I like the idea, but this is broken and will cause desyncs as written. :-1:

**Comment by Phrohdoh** [comment]
Oh right, I'll rewrite with that in mind.

**Comment by Phrohdoh** [comment]
Updated, @chrisforbes it should be sync-safe now.

**Comment by Phrohdoh** [comment]
Updated again.
Sync'd state rules are fickle.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 2]
```diff
@@ -72,9 +72,21 @@ IEnumerable<Order> InnerOrder(World world, CPos xy, MouseInput mi)
...
 				var topLeft = xy - FootprintUtils.AdjustForBuildingSize(buildingInfo);
-				if (!world.CanPlaceBuilding(building, buildingInfo, topLeft, null)
-					|| !buildingInfo.IsCloseEnoughToBase(world, producer.Owner, building, topLeft))
+
+				var closeEnough = buildingInfo.IsCloseEnoughToBase(world, producer.Owner, building, topLeft);
+				var canPlace = world.CanPlaceBuilding(building, buildingInfo, topLeft, null);
+
+				if (!closeEnough || !canPlace)
 				{
+					if (!canPlace)
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Should this block run when `closeEnough` is false?

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 3]
```diff
@@ -72,9 +72,21 @@ IEnumerable<Order> InnerOrder(World world, CPos xy, MouseInput mi)
 			if (mi.Button == MouseButton.Left)
 			{
 				var topLeft = xy - FootprintUtils.AdjustForBuildingSize(buildingInfo);
-				if (!world.CanPlaceBuilding(building, buildingInfo, topLeft, null)
-					|| !buildingInfo.IsCloseEnoughToBase(world, producer.Owner, building, topLeft))
+
+				var closeEnough = buildingInfo.IsCloseEnoughToBase(world, producer.Owner, building, topLeft);
+				var canPlace = world.CanPlaceBuilding(building, buildingInfo, topLeft, null);
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: if `closeEnough` is false here, you can skip this and save a few cycles.

**Comment by obrakmann** [comment]
I get this each time I use this feature (in multiplayer only, not in singleplayer):
screenshot from 2015-04-05 23 01 41

**Comment by Phrohdoh** [comment]
@obrakmann Interesting message, unfortunately I will have to rework this completely.

**Comment by chrisforbes** [comment]
This will be caused by issuing orders from synced code. Only the owning
client is allowed to issue orders to a unit.

**Comment by Phrohdoh** [comment]
Closing for now.
`PlaceBuilding` is a mess and a headache. I also can't seem to figure out valid sync code.

**Comment by x-a-n-a-x** [comment]
I notice that you deleted your branch, Phrohdoh...
Does that mean you don't plan to work anymore on this feature ? :-(

**Comment by MicroJOo** [comment]
i also noticed that you deleted your branch @Phrohdoh ... but i would like to have that feature ... i think once your headache is gone its worth giving it another try ...

**Comment by WolfGaming** [comment]
putting this on my will have a look at when I have the time list.

**Pull request closed without merging**